common/listener.go

```diff
@@ -41,142 +45,188 @@ func NewListener(u *url.URL) (*Listener, error) {
 }
 
-func GetTlsConfig(tlsSettings localtls.TlsSettings) (*tls.Config, error) {
+func GetTLSConfig(tlsSettings localtls.TlsSettings, errChan chan<- error) (*tls.Config, error) {
+	if tlsSettings.TLSCertificatePath != "" {
+		return getTLSConfigFromFile(tlsSettings)
+	}
+
+	if tlsSettings.KeyVaultURL != "" {
+		return getTLSConfigFromKeyVault(tlsSettings, errChan)
+	}
+
+	return nil, errors.Errorf("invalid tls settings: %+v", tlsSettings)
+}
+
+func getTLSConfigFromFile(tlsSettings localtls.TlsSettings) (*tls.Config, error) {
 	tlsCertRetriever, err := localtls.GetTlsCertificateRetriever(tlsSettings)
 	if err != nil {
-		return nil, fmt.Errorf("Failed to get certificate retriever %+v", err)
+		return nil, errors.Wrap(err, "failed to get certificate retriever")
 	}
+
 	leafCertificate, err := tlsCertRetriever.GetCertificate()
 	if err != nil {
-		return nil, fmt.Errorf("Failed to get certificate %+v", err)
+		return nil, errors.Wrap(err, "failed to get certificate")
 	}
+
 	if leafCertificate == nil {
-		return nil, fmt.Errorf("Certificate retrival returned empty %+v", err)
+		return nil, errors.New("certificate retrieval returned empty")
 	}
+
 	privateKey, err := tlsCertRetriever.GetPrivateKey()
 	if err != nil {
-		return nil, fmt.Errorf("Failed to get certificate private key %+v", err)
+		return nil, errors.Wrap(err, "failed to get certificate private key")
 	}
+
 	tlsCert := tls.Certificate{
 		Certificate: [][]byte{leafCertificate.Raw},
 		PrivateKey:  privateKey,
 		Leaf:        leafCertificate,
 	}
+
 	tlsConfig := &tls.Config{
-		MaxVersion: tls.VersionTLS12,
+		MaxVersion: tls.VersionTLS13,
 		MinVersion: tls.VersionTLS12,
 		Certificates: []tls.Certificate{
 			tlsCert,
 		},
 	}
+
 	return tlsConfig, nil
 }
 
-// Start creates the listener socket and starts the HTTPS server.
-func (listener *Listener) StartTLS(errChan chan<- error, tlsSettings localtls.TlsSettings) error {
-	tlsConfig, err := GetTlsConfig(tlsSettings)
+func getTLSConfigFromKeyVault(tlsSettings localtls.TlsSettings, errChan chan<- error) (*tls.Config, error) {
+	credOpts := azidentity.ManagedIdentityCredentialOptions{ID: azidentity.ResourceID(tlsSettings.MSIResourceID)}
+	cred, err := azidentity.NewManagedIdentityCredential(&credOpts)
 	if err != nil {
-		log.Printf("[Listener] Failed to compose Tls Configuration with errror: %+v", err)
-		return err
+		return nil, errors.Wrap(err, "could not create managed identity credential")
 	}
+
+	kvs, err := keyvault.NewShim(tlsSettings.KeyVaultURL, cred)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not create new keyvault shim")
+	}
+
+	// todo: need to determine if we can use zapai
+	l, _ := zap.NewProduction()
+	ctx := context.TODO()
+
+	cr, err := keyvault.NewCertRefresher(ctx, kvs, l, tlsSettings.KeyVaultCertificateName)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not create new cert refresher")
+	}
+
+	go func() {
+		errChan <- cr.Refresh(ctx, tlsSettings.KeyVaultCertificateRefreshInterval)
```

Comment:
I don't think this will ever return cleanly unless the certificate's expiration is before the next refresh interval. The context has no cancellation. That might be fine if it's intended to share the lifetime of the application, but I figured I'd mention it.

Reply:
Yeah, ideally, context would be propagated all the way from `main` to handle signals cleanly but there are a few layers of indirection unfortunately, so I opted for a `context.TODO()` in the meantime. This goroutine is indeed tied to the lifetime of the application so a temporary leak _might_ be ok for now. Since I have some changes to push, I may try to propagate context, but will depend on size of changes.